auth/scram.go

```diff
@@ -0,0 +1,253 @@
+	hm.Write(buff)
+	return hm.Sum(nil)
+}
+
+func (s *ScramManager) lookupCredential(username string) (scram.StoredCredentials, error) {
+	batch, err := s.executeQuerySingleResultBatch(GetCredsQueryName, []any{username})
+	if err != nil {
+		return scram.StoredCredentials{}, err
+	}
+	user := batch.Columns[0].(*evbatch.StringColumn).Get(0)
+	if user != username {
+		// sanity check
+		return scram.StoredCredentials{}, errors.New("returned username does not match requested username")
+	}
+	credsBuff := batch.Columns[1].(*evbatch.BytesColumn).Get(0)
+	var creds userCreds
+	if err := json.Unmarshal(credsBuff, &creds); err != nil {
+		return scram.StoredCredentials{}, err
+	}
+	var storedCreds scram.StoredCredentials
+	v, err := base64.StdEncoding.DecodeString(creds.StoredKey)
+	if err != nil {
+		return scram.StoredCredentials{}, err
+	}
+	storedCreds.StoredKey = v
+	v, err = base64.StdEncoding.DecodeString(creds.ServerKey)
+	if err != nil {
+		return scram.StoredCredentials{}, err
+	}
+	storedCreds.ServerKey = v
+	storedCreds.KeyFactors.Iters = creds.Iters
+	storedCreds.KeyFactors.Salt = creds.Salt
+	return storedCreds, nil
+}
+
+func (s *ScramManager) executeQuerySingleResultBatch(queryName string, args []any) (*evbatch.Batch, error) {
+	ch := make(chan *evbatch.Batch, 1)
+	_, err := s.queryManager.ExecutePreparedQueryWithHighestVersion(queryName, args, math.MaxInt64,
+		func(last bool, numLastBatches int, batch *evbatch.Batch) error {
+			if numLastBatches != 1 {
+				panic("sys query must have 1 partition")
+			}
+			if !last {
+				panic("must be one result batch")
+			}
+			ch <- batch
+			return nil
+		})
+	if err != nil {
+		return nil, err
+	}
+	return <-ch, nil
+}
+
+func (s *ScramManager) sendCredsBatch(batch *proc.ProcessBatch) error {
+	ch := make(chan error, 1)
+	// We ingest this with replication, so the batch will not be lost if failure occurs.
+	s.forwarder.ForwardBatch(batch, true, func(err error) {
+		ch <- err
+	})
+	return <-ch
+}
+
+func (s *ScramManager) NewConversation(scramAuthType ScramAuthType) (*ScramConversion, error) {
+	var conv *scram.ServerConversation
+	if scramAuthType == ScramAuthTypeSHA256 {
+		conv = s.scramServerSha256.NewConversation()
+	} else if scramAuthType == ScramAuthTypeSHA512 {
+		conv = s.scramServerSha512.NewConversation()
+	} else {
+		panic("unknown scram auth type")
+	}
+	return &ScramConversion{
+		conv: conv,
+	}, nil
+}
+
+type ScramConversion struct {
+	principal string
+	conv      *scram.ServerConversation
+	lock      sync.Mutex
+}
+
+func (s *ScramConversion) Principal() string {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	return s.principal
+}
+
+func (s *ScramConversion) Process(request []byte) (resp []byte, complete bool, failed bool) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	r, err := s.conv.Step(string(request))
+	if err != nil {
+		// Log auth failures at info
+		log.Infof("Kafka API SASL SCRAM authentication failure: %v", err)
+		return nil, false, true
+	}
+	if s.conv.Done() {
```

Comment:
Should this be `s.conv.Valid()` ? Reading the `ServerConversation` docs, it says that 
```
Valid returns true if the conversation successfully authenticated the client.
```

Reply:
After digging in `xdg-go/scram` a bit, I can see that `Step()` will only return a nil error if the authentication succeeded, but I feel relying on that makes the code assume too much knowledge and risks breaking in case that changes. I think the design is made for `Valid` and `Done` to have different purposes.

Reply:
Yes, good catch.. although (and I think you made the same point later on), the conv returns an error if not valid, so it does work anyway. But yes, we should check this too.